### PR TITLE
Add UUID and GUID predicates

### PR DESCRIPTION
**Pull request by gertjanal**
Constants for Microsoft GUID and Java UUID, to be used in predicates

### CODE REVIEW

File: formats/src/main/java/io/parsingdata/metal/format/UID.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    public static ValueExpression guid(final String guid) {
+        final String[] parts = guid.split("-");
+        if (parts.length != 5) {
+            throw new IllegalArgumentException("Invalid GUID string: " + guid);
+        }
+
+        // Note that GUID bytes differ from UUID bytes, as the first 3 parts are reversed
+        // Use ByteBuffer instead of long to make sure no leading zeroes are omitted
+        final ByteBuffer buffer = ByteBuffer.allocate(16);
+        buffer.putInt(0, Integer.reverseBytes((int) Long.parseLong(parts[0], 16)));
```
Comment (rdvdijk) [inline formats/src/main/java/io/parsingdata/metal/format/UID.java]: The endianness of the first three are defined as "Native". This means that we need to be able to define the encoding of those fields when we create a `guid` value expression.

See see *Binary encoding* section in the [Wikipedia GUID article](https://en.wikipedia.org/wiki/Globally_unique_identifier).

Comment (gertjanal) [inline formats/src/main/java/io/parsingdata/metal/format/UID.java]: Added fix

File: formats/src/main/java/io/parsingdata/metal/format/UID.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * Use a String representation of a GUID as predicate.
+     * {@code eq(guid("caa16737-fa36-4d43-b3b6-33f0aa44e76b"))}
+     * Note that the byte order in the encoding matters for the output.
+     * @param guid GUID, for example "caa16737-fa36-4d43-b3b6-33f0aa44e76b"
+     * @return expression to use as predicate
+     */
+    public static ValueExpression guid(final String guid) {
+        final String[] parts = guid.split("-");
```
Comment (rdvdijk) [inline formats/src/main/java/io/parsingdata/metal/format/UID.java]: `checkNotNull`?

File: core/src/main/java/io/parsingdata/metal/expression/derived/GUID.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        }
+        return new ValueExpression() {
+
+            @Override
+            public OptionalValueList eval(final Environment environment, final Encoding encoding) {
+                // Note that GUID bytes differ from UUID bytes, as the first 3 parts can be reversed
+                return cat(
+                    cat(
+                        cat(
+                            four(parts[0], encoding),
```
Comment (gertjanal) [inline core/src/main/java/io/parsingdata/metal/expression/derived/GUID.java]: @rdvdijk cool huh? :D

Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/expression/derived/GUID.java]: I like :+1:

Comment (ccreeten) [inline core/src/main/java/io/parsingdata/metal/expression/derived/GUID.java]: Wouldn't depend on the default settings of Encoding, at least not until it is specifically stated :laughing:

File: core/src/test/java/io/parsingdata/metal/expression/derived/UUIDTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+public class UUIDTest {
+
+    @Test
+    public void parseUUID() throws IOException, URISyntaxException {
+        final Token uuid = def("uuid", 16, eq(uuid("00c27766-f623-4200-9d64-115e9bfd4a08")));
+        assertTrue(uuid.parse(stream(0x00, 0xc2, 0x77, 0x66, 0xf6, 0x23, 0x42, 0x00, 0x9d, 0x64, 0x11, 0x5e, 0x9b, 0xfd, 0x4a, 0x08), enc()).succeeded);
+        assertFalse(uuid.parse(stream(0x66, 0x77, 0xc2, 0x00, 0x23, 0xf6, 0x00, 0x42, 0x9d, 0x64, 0x11, 0x5e, 0x9b, 0xfd, 0x4a, 0x08), enc()).succeeded);
+
+        new UUID(); // test coverage
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/expression/derived/UUIDTest.java]: `core` has a `checkUtilityClass()` utility function, check it out.

File: core/src/main/java/io/parsingdata/metal/expression/derived/UUID.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * {@link UUID#uuid(String)} creates a ValueExpression to be used as predicate for 16 byte definitions;
+ *
+ * @author Netherlands Forensic Institute.
+ */
+public final class UUID {
+    private static final Encoding BIG_ENDIAN = new Encoding();
+
+    UUID() {}
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/expression/derived/UUID.java]: (See my `checkUtilityClass` comment.) Should be `private`.

File: core/src/main/java/io/parsingdata/metal/Shorthand.java [hunk 6]
```diff
@@ -175,6 +177,9 @@ private Shorthand() {}
     public static ComparisonExpression ltNum(final ValueExpression predicate) { return new LtNum(null, predicate); }
     public static ComparisonExpression ltNum(final ValueExpression value, final ValueExpression predicate) { return new LtNum(value, predicate); }
 
+    public static ValueExpression guid(final String guid) {return GUID.guid(guid); }
+    public static ValueExpression uuid(final String uuid) {return UUID.uuid(uuid); }
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/Shorthand.java]: Bonus nitpick comment: Add space after `{`.

File: core/src/test/java/io/parsingdata/metal/expression/derived/UUIDTest.java (hunk 4 above)
Comment (gertjanal) [inline core/src/test/java/io/parsingdata/metal/expression/derived/UUIDTest.java]: thnx, added

File: core/src/main/java/io/parsingdata/metal/expression/derived/UUID.java (hunk 5 above)
Comment (gertjanal) [inline core/src/main/java/io/parsingdata/metal/expression/derived/UUID.java]: done

File: core/src/main/java/io/parsingdata/metal/Shorthand.java (hunk 6 above)
Comment (gertjanal) [inline core/src/main/java/io/parsingdata/metal/Shorthand.java]: added

**Comment by rdvdijk** [comment]
@jvdb Merge before next release?

**Pull request merged**